### PR TITLE
add Game#log_later method, log phase change after train purchase

### LINKED ISSUE #332: Change order of history messages for train purchase and phase change
"-- Phase 3 (Operating Rounds: 2, Train Limit: 4, Available Tiles: Yellow, Green ) --"
"KO buys a 3 train for ¥180 from The Depot"

I think these messages should be in the opposite order. The phase changes immediately *after* the train is bought, not before.

### PR DESCRIPTION
**Pull request by michaeljb**
[Fixes #332]

### CODE REVIEW

**Comment by perwestling** [comment]
Wonder how it logs in 18MEX, when 5 train triggers NdM merge?

**Comment by michaeljb** [comment]
Looks like it handles the NdM merge nicely. I used https://18xx.games/game/15697 for this example.

Screenshots showing how the log looks partway through the merge process:

## Current master behavior

Screenshot from 2020-11-20 08-38-53

## New behavior

Screenshot from 2020-11-20 08-39-08

Copy-paste showing how the log looks at the very end of the NdM merge process:

## Current master behavior

SPM does not run
SPM's share price changes from $75 to $70
**-- Phase 5 (Operating Rounds: 3, Train Limit: 2, Available Tiles: Yellow, Green, Brown ) --**
**-- Event: Private companies close --**
**-- Event: NdM merger --**
Merge candidates: MEX (chili), UdY (DanielTroi), CHI (DanielTroi), SPM (secondcousinstanley), MC (secondcousinstanley)
SPM buys a 5 train for $450 from The Depot
MEX declines
UdY declines
CHI declines
SPM declines
**-- MC merges into NdM --**
Swedish-Per (GMT+2) receives $15 in share compensation
DanielTroi receives $15 in share compensation
secondcousinstanley receives a 10% share of NdM
chili receives $15 in share compensation
MC's token in I8 is replaced with an NdM token
NdM receives the MC treasury of $41
NdM receives the trains: 3,3
SPM buys a 5 train for $450 from The Depot


## New behavior on this branch

SPM does not run
SPM's share price changes from $75 to $70
SPM buys a 5 train for $450 from The Depot
**-- Phase 5 (Operating Rounds: 3, Train Limit: 2, Available Tiles: Yellow, Green, Brown ) --**
**-- Event: Private companies close --**
**-- Event: NdM merger --**
Merge candidates: MEX (chili), UdY (DanielTroi), CHI (DanielTroi), SPM (secondcousinstanley), MC (secondcousinstanley)
MEX declines
UdY declines
CHI declines
SPM declines
**-- MC merges into NdM --**
Swedish-Per (GMT+2) receives $15 in share compensation
DanielTroi receives $15 in share compensation
secondcousinstanley receives a 10% share of NdM
chili receives $15 in share compensation
MC's token in I8 is replaced with an NdM token
NdM receives the MC treasury of $41
NdM receives the trains: 3,3
SPM buys a 5 train for $450 from The Depot

File: lib/engine/game/base.rb [hunk 1]
```diff
@@ -1009,6 +1010,17 @@ def after_par(corporation)
         end
       end
 
+      def log_later(&block)
```
Comment (tobymao) [inline lib/engine/game/base.rb:1013]: looks like this only works if called once with a matching log_now, could we make it more robust?

also no need to pass in a block, you can just use yield

Comment (michaeljb) [reply lib/engine/game/base.rb]: > looks like this only works if called once with a matching log_now, could we make it more robust?

9bc9ba3a has the game return the lines without storing them in `@log_later`; this still only works if something logs the lines later, but in the caller's code it's more obvious now that the caller should do something with the lines. This is also closer to the `log_later` variable that is used in `step/buy_company.rb`

>  no need to pass in a block, you can just use yield

452744bc

Comment (tobymao) [reply lib/engine/game/base.rb]: i think i prefered the previous version maybe rename it

queue_log

flush_log

??

Comment (michaeljb) [reply lib/engine/game/base.rb]: Renamed and rebased

**Pull request merged**